### ISSUE TITLE
replaceOnConflict: true is not respected for saving links

### ISSUE THREAD

**Issue by leisim** [isar/isar#354]
I ran this code:

```
String json = '{"myServerId": "id123", "name": "Jhon", "teacher": {"mySeerverId": "id1234", "subject": "Math"}}';

    await isar!.writeTxn((isar) async {
      for(int i = 0; i < 10000; i++) {
        Student student = studentFromJson(json);
        student.name = student.name! +' '+i.toString();
        await isar.students.put(student, replaceOnConflict: true, saveLinks: true);
      }
    });


    var count = await isar!.teachers.countSync();
    print(count);

```
and I got an error: `[VERBOSE-2:ui_dart_state.cc(209)] Unhandled Exception: IsarError: Unique index violated.`

But when I run without `saveLinks: true` I get no errors and don't save (of course) teacher, its look like `replaceOnConflict: true` affect only the main `@Collection()`

_Originally posted by @gustavorozolin in https://github.com/isar/isar/discussions/344#discussioncomment-2639003_

**Comment by flocbit** [comment]
Seems like this is bigger than just replaceOnConflict.

Since 2.4.0 when I do:

```
collection.locations.add(
      model.toIsar(),
);
```

and afterwards

```
db.writeTxnSync(
     (isar) => collection.locations.saveSync(),
);
```

nothing happens.

I have checked `collection.locations` and right before I try to save it the new entry exists in the temporary array. So I assume the save process does not work.

**Comment by flocbit** [comment]
I can also confirm that by reverting to `2.2.1` everything works just fine.

**Comment by leisim** [comment]
Currently I would have to write all tests twice to cover sync methods and they are therefore not tested well.

I'm working on a fix to test sync and async methods simultaneously.

**Comment by gustavorozolin** [comment]
@flocbit I downgrade to 2.2.1 version and got the same error:

```
 String json =
        '{"myServerId": "id123", "name": "John", "teacher": {"mySeerverId": "id1234", "subject": "Math"}}';

await isar!.writeTxn((isar) async {
      for (int i = 0; i < 10000; i++) {
        print(i.toString());
        Student student = studentFromJson(json);
        student.name = student.name! + ' ' + i.toString();
        student.myServerId = student.myServerId! + ' ' + i.toString();
        await isar.students.put(student, replaceOnConflict: true, saveLinks: true);
      }
    });
```

Error: Unhandled Exception: IsarError: Unique index violated.

**Comment by leisim** [comment]
This is fixed in `3.0.0-dev.0`. You can now decide for each index how conflicts should be handled 🙂
